src/torchjd/_autogram/_rev_gram_acc.py

```diff
@@ -0,0 +1,287 @@
+        self.counter.update(tensors)
+
+    def add_jacobian(self, tensor: Tensor, jacobian: Tensor) -> None:
+        if tensor in self.jacobians:
+            self.jacobians[tensor] += jacobian
+        else:
+            self.jacobians[tensor] = jacobian
+        self.counter.subtract([tensor])
+        if self.counter[tensor] == 0:
+            self._accumulate_jacobian(self.jacobians[tensor])
+            del self.counter[tensor]
+            del self.jacobians[tensor]
+
+    def _accumulate_jacobian(self, jacobian: Tensor) -> None:
+        jacobian_matrix = torch.flatten(jacobian, start_dim=1)
+        if self._gramian is not None:
+            self._gramian.addmm_(jacobian_matrix, jacobian_matrix.T)
+        else:
+            self._gramian = torch.mm(jacobian_matrix, jacobian_matrix.T)
+
+    @property
+    def gramian(self) -> Tensor:
+        if len(self.counter) != 0 or len(self.jacobians) != 0:
+            shape_to_count = [(k.shape, v) for k, v in self.counter.items()]
+            raise ValueError(
+                f"Some tracked parameters are still not at a count of 0, {shape_to_count}."
+            )
+
+        return self._gramian
+
+
+def make_jacobian_accumulator(
+    module: nn.Module,
+    gramian_accumulator: GramianAccumulator,
+    args: PyTree,
+    tree_spec: TreeSpec,
+) -> type[torch.autograd.Function]:
+
+    activated = True
+
+    class JacobianAccumulator(torch.autograd.Function):
+        @staticmethod
+        def forward(*xs: Tensor) -> tuple[Tensor, ...]:
+            return tuple([x.detach() for x in xs])
+
+        @staticmethod
+        def setup_context(*_):
+            pass
+
+        @staticmethod
+        def backward(ctx, *flat_grad_outputs: Tensor):
+            nonlocal activated
+            if activated:
+
+                def get_vjp(grad_outputs_j: PyTree, inputs_j: PyTree) -> tuple[Tensor, ...]:
+                    # Note: we use unsqueeze(0) to turn a single activation (or grad_output) into a
+                    # "batch" of 1 activation (or grad_output). This is because some layers (e.g.
+                    # nn.Flatten) do not work equivalently if they're provided with a batch or with
+                    # an element of a batch. We thus always provide them with batches, just of a
+                    # different size.
+                    inputs_j = tree_map(lambda x: x.unsqueeze(0), inputs_j)
+                    grad_outputs_j = tree_map(lambda x: x.unsqueeze(0), grad_outputs_j)
+
+                    # _vjp_from_module returns a function that computes the vjp w.r.t. to the
+                    # primals (tuple), here the functional has a single primal which is
+                    # dict(module.named_parameters()). We therefore take the 0'th element to obtain
+                    # the dict of gradients w.r.t. the module's named_parameters.
+                    return _vjp_from_module(module, *inputs_j)(grad_outputs_j)[0]
```

Comment:
```suggestion
                    return _vjp_from_module(module, inputs_j)(grad_outputs_j)[0]
```
I think the curent version is wrong unless `inputs_j` is a tuple, which it seems to be. But in any case, I think that `_vjp_from_module` could (and should) take a `PyTree`

Reply:
Good catch! In my recent test (`PyTreeInputPyTreeOutputModule`), `inputs` is a list, so unpacking works, but in general you're right that it should not work.